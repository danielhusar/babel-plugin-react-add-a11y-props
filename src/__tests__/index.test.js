const babel = require('@babel/core');
const reactPreset = require('@babel/preset-react');
const prettier = require('prettier');
const convertReactCreateElementToJSX = require('babel-plugin-transform-react-createelement-to-jsx');
const plugin = require('../');

const format = jsx => {
  const { code: reactElements } = babel.transform(jsx, { plugins: [plugin], presets: [reactPreset] });
  const { code } = babel.transform(reactElements, { plugins: [convertReactCreateElementToJSX] });
  return prettier.format(code, { printWidth: 300, parser: 'babylon' });
};

// babel-plugin-transform-react-createelement-to-jsx have some deprecated calls, so silent them until it upgrades
console.error = jest.fn();

describe('react-add-a11y-props', () => {
  it('add onKeyDown, tabindex and role to all elements with onClick', () => {
    const jsx = `
      <span>
        <span onClick={cb} />
        <div onClick={cb} />
        <article onClick={cb} />
        <a onClick={() => foo()} />
      </span>
    `;
    expect(format(jsx)).toMatchSnapshot();
  });

  it('it will only add tabindex to inputs and will not add keyDown and role', () => {
    const jsx = `
      <span>
        <input type="submit" onClick={cb} />
        <button onClick={cb} />
        <select onClick={() => foo()} />
      </span>
    `;
    expect(format(jsx)).toMatchSnapshot();
  });

  it('it will not add tabIndex and keyDown for negative tabIndex', () => {
    const jsx = `<div tabIndex="-1" onClick={cb} />`;
    expect(format(jsx)).toMatchSnapshot();
  });

  it('it will not add tabIndex if its already present', () => {
    const jsx = `<div tabIndex="1" onClick={cb} />`;
    expect(format(jsx)).toMatchSnapshot();
  });

  it('it will not add role if its already present', () => {
    const jsx = `<div role="foo" onClick={cb} />`;
    expect(format(jsx)).toMatchSnapshot();
  });

  it('it will not add keyDown if its already present', () => {
    const jsx = `<div onKeyDown={cb} onClick={cb} />`;
    expect(format(jsx)).toMatchSnapshot();
  });

  it('it will not add keyDown and role to anchor with href', () => {
    const jsx = `<a href="https://www.google.com" onClick={cb} />`;
    expect(format(jsx)).toMatchSnapshot();
  });
});
