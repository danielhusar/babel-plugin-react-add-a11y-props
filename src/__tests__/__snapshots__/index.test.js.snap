// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-add-a11y-props add onKeyDown, tabindex and role to all elements with onClick 1`] = `
"<span>
  <span onClick={cb} onKeyDown={e => (e.keyCode === 13 || e.keyCode === 32) && cb(e)} role=\\"button\\" tabIndex=\\"0\\" />
  <div onClick={cb} onKeyDown={e => (e.keyCode === 13 || e.keyCode === 32) && cb(e)} role=\\"button\\" tabIndex=\\"0\\" />
  <article onClick={cb} onKeyDown={e => (e.keyCode === 13 || e.keyCode === 32) && cb(e)} role=\\"button\\" tabIndex=\\"0\\" />
  <a onClick={() => foo()} onKeyDown={e => (e.keyCode === 13 || e.keyCode === 32) && (() => foo())(e)} role=\\"button\\" tabIndex=\\"0\\" />
</span>;
"
`;

exports[`react-add-a11y-props it will not add keyDown and role to anchor with href 1`] = `
"<a href=\\"https://www.google.com\\" onClick={cb} tabIndex=\\"0\\" />;
"
`;

exports[`react-add-a11y-props it will not add keyDown if its already present 1`] = `
"<div onKeyDown={cb} onClick={cb} role=\\"button\\" tabIndex=\\"0\\" />;
"
`;

exports[`react-add-a11y-props it will not add role if its already present 1`] = `
"<div role=\\"foo\\" onClick={cb} onKeyDown={e => (e.keyCode === 13 || e.keyCode === 32) && cb(e)} tabIndex=\\"0\\" />;
"
`;

exports[`react-add-a11y-props it will not add tabIndex and keyDown for negative tabIndex 1`] = `
"<div tabIndex=\\"-1\\" onClick={cb} />;
"
`;

exports[`react-add-a11y-props it will not add tabIndex if its already present 1`] = `
"<div tabIndex=\\"1\\" onClick={cb} onKeyDown={e => (e.keyCode === 13 || e.keyCode === 32) && cb(e)} role=\\"button\\" />;
"
`;

exports[`react-add-a11y-props it will only add tabindex to inputs and will not add keyDown and role 1`] = `
"<span>
  <input type=\\"submit\\" onClick={cb} tabIndex=\\"0\\" />
  <button onClick={cb} tabIndex=\\"0\\" />
  <select onClick={() => foo()} tabIndex=\\"0\\" />
</span>;
"
`;
